### PR TITLE
*: simplify raft cluster.

### PR DESCRIPTION
**Pull request by siddontang**
For #68 

@ngaut @qiuyesuifeng @disksing

### CODE REVIEW

File: server/cluster_test.go [hunk 1]
```diff
@@ -297,15 +307,42 @@ func (s *testClusterSuite) TestGetPutMeta(c *C) {
...
+	testClusterBaseSuite
+}
+
+func (s *testClusterCacheSuite) getRootPath() string {
+	return "test_cluster_cache"
+}
+
+func (s *testClusterCacheSuite) SetUpSuite(c *C) {
+	s.svr = newTestServer(c, s.getRootPath())
+
```
Comment (ngaut) [inline server/cluster_test.go]: Remove empty line.

File: server/cluster_test.go [hunk 2]
```diff
@@ -297,15 +307,42 @@ func (s *testClusterSuite) TestGetPutMeta(c *C) {
...
+	s.client = newEtcdClient(c)
+
```
Comment (ngaut) [inline server/cluster_test.go]: Ditto.

File: server/cluster_test.go [hunk 3]
```diff
@@ -297,15 +307,42 @@ func (s *testClusterSuite) TestGetPutMeta(c *C) {
...
+}
+
+func (s *testClusterCacheSuite) TearDownSuite(c *C) {
+	s.svr.Close()
+	s.client.Close()
+}
+
+func (s *testClusterCacheSuite) TestCache(c *C) {
+	mustGetLeader(c, s.client, s.svr.getLeaderPath())
+
```
Comment (ngaut) [inline server/cluster_test.go]: Ditto

File: server/cluster_test.go [hunk 4]
```diff
@@ -297,15 +307,42 @@ func (s *testClusterSuite) TestGetPutMeta(c *C) {
...
+	clusterID := uint64(0)
 
 	req := s.newBootstrapRequest(c, clusterID, "127.0.0.1:1")
 	store1 := req.Bootstrap.Store
 
-	s.svr.bootstrapCluster(clusterID, req.Bootstrap)
+	s.svr.bootstrapCluster(req.Bootstrap)
```
Comment (ngaut) [inline server/cluster_test.go]: Ditto.

File: server/cluster_test.go (hunk 3 above)
Comment (siddontang) [inline server/cluster_test.go]: should use an empty line to separate here.

File: server/cluster.go [hunk 5]
```diff
@@ -28,24 +28,28 @@ var (
 )
 
 // Raft cluster key format:
-// cluster 1 -> /raft/1, value is metapb.Cluster
-// cluster 2 -> /raft/2
+// cluster 1 -> /1/raft, value is metapb.Cluster
+// cluster 2 -> /1/raft
```
Comment (qiuyesuifeng) [inline server/cluster.go]: The comment should be `/2/raft`?

Comment (siddontang) [inline server/cluster.go]: yep

File: server/cluster.go [hunk 6]
```diff
@@ -28,24 +28,28 @@ var (
...
-// region 1 -> /raft/1/r/1, value is encoded_region_key
-// region search key map -> /raft/1/k/encoded_region_key, value is metapb.Region
+// store 1 -> /1/raft/s/1, value is metapb.Store
+// region 1 -> /1/raft/r/1, value is encoded_region_key
+// region search key map -> /1/raft/k/encoded_region_key, value is metapb.Region
 //
 // Operation queue, pd can only handle operations like auto-balance, split,
 // merge sequentially, and every operation will be assigned a unique incremental ID.
-// pending queue -> /raft/1/j/1, /raft/1/j/2, value is operation job.
+// pending queue -> /1/raft/j/1, /raft/1/j/2, value is operation job.
```
Comment (qiuyesuifeng) [inline server/cluster.go]: `/1/raft/j/2`

**Comment by qiuyesuifeng** [comment]
LGTM

File: pd-server/main.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ var (
 	leaderLease   = flag.Int64("lease", 3, "Leader lease time (second)")
 	logLevel      = flag.String("L", "debug", "log level: info, debug, warn, error, fatal")
 	pprofAddr     = flag.String("pprof", ":6060", "pprof HTTP listening address")
+	clusterID     = flag.Uint64("cluster-id", 0, "Cluster ID")
```
Comment (disksing) [inline pd-server/main.go]: Maybe another default clusterID since 0 is for test.

File: server/cluster.go [hunk 8]
```diff
@@ -69,16 +73,21 @@ type raftCluster struct {
...
-		clusterRoot: s.getClusterRootPath(clusterID),
-		askJobCh:    make(chan struct{}, askJobChannelSize),
-		quitCh:      make(chan struct{}),
-		storeConns:  newStoreConns(defaultConnFunc),
+func (c *raftCluster) Start(meta metapb.Cluster) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.running {
+		log.Warn("raft cluster is already stared")
```
Comment (disksing) [inline server/cluster.go]: "raft cluster has already started"

File: pd-server/main.go (hunk 7 above)
Comment (siddontang) [inline pd-server/main.go]: check 0 and panic?

Comment (disksing) [inline pd-server/main.go]: I think make the default value to be 1 is enough.

Comment (siddontang) [inline pd-server/main.go]: em, 1 seems strange, I prefer 0 if 1. 

We can log a warning if we use 0. Btw for most cases, we use different root path not "/pd", so it is ok.

Comment (disksing) [inline pd-server/main.go]: Fine.

**Comment by disksing** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
LGTM

**Pull request merged**